File: hash-tables/two-sum/go/two_sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	complements := make(map[int]int)
	ans := []int{}

	for index, val := range nums {
		complement := target - val
		if cIndex, ok := complements[complement]; ok {
			ans = []int{index, cIndex}
		}
		complements[val] = index
	}

	return ans
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9

	ans := twoSum(nums, target)

	fmt.Printf("%v\n", ans)
}

File: hash-tables/two-sum/python/two_sum.py
# https://leetcode.com/problems/two-sum/

from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        complements = {}
        ans = []

        for index, val in enumerate(nums):
            complement = target - val

            if complement in complements:
                ans = [index, complements[complement]]
                break

            complements[val] = index

        return ans


def main():
    nums = [2, 7, 11, 15]
    target = 9

    print(Solution().twoSum(nums, target))


if __name__ == "__main__":
    main()
